### PR TITLE
1091 - DNSSEC UI with Key Data

### LINKED ISSUE #1091: UI Dev: Update DNSSEC
### Issue Description

Story
As a domain manager OR CISA Analyst,
I want update DNSSEC data for a particular domain
so that the authenticity of the provided DNS data may be verified.

Acceptance Criteria
- [ ]  Implement design created in https://github.com/cisagov/getgov/issues/893
 specific to DNSSEC.
- [ ] In the event of an error, display to the user an appropriate message for the error. However, this should be as basic as possible for use in November. Including:
       - showing Registry Errors when they occur. 
       - The front end should validate the data types and show errors on the given fields that violate these type restrictions
       - out of scope: showing which fields resulted in registry errors and making the registry errors less technical

### Additional Context (optional)

_No response_

### Issue Links

Depends on https://github.com/cisagov/getgov/issues/893 (UI Design)
Split from #1016

### PR DESCRIPTION
**Pull request by dave-kennedy-ecs**
## Ticket

Resolves #1091 #1122

## Changes

- DNS page hierarchy in Domain Detail, including name servers and dnssec
- DNSSEC page which allows for enabling and disabling DNSSEC, and adding ds data and key data
- Left Sidebar content depends on session state and availability of DNSSEC data in the domain
- DS Data Record page with form for editing DS Data Records
  - form includes repeatable fieldsets, one for each DS Data record
  - fieldsets/records can be deleted
  - field level and form level error handling
  - cancel to undo changes before they are saved to the registry
  - form updates registry upon save
- DS Key Record page with form for editing DS Key Records
  - form includes repeatable fieldsets, one for each DS Key Data record
  - fieldsets/records can be deleted
  - field level and form level error handling
  - cancel to undo changes before they are saved to the registry
  - form updates registry upon save (although registry currently erroring on Key Data)
- Modal on DNSSEC page for Disable DNSSEC
  - Disable DNSSEC removes all dnssecdata from registry and disables DNSSEC
- Refactor of SCSS to make files more modular (does not effect the css output)
- Refactored all of the test_views test cases
- DNSSEC setter now makes two EPP calls instead of one. First call adds new DNSSEC data. Second call deletes DNSSEC data. These calls are determined by comparing existing DNSSEC data with new DNSSEC data being set.

## Context for reviewers

DNSSEC page:

When you first land on the DNSSEC page you will see a "Enable DNSSEC" button. When you click that button, a session var is set and now you see buttons to Add DS or Add Key data, plus a cancel that toggles that session var.

If you navigate away and come back without having added any data, you land back on "Enable DNSSEC"

If you add or already have data added for that domain, you will see a "Disable DNSSEC" button instead. Clicking that button will prompt a modal where you can conform or cancel. If you click confirm in the modal, the data will be deleted and you are back to "Enable DNSSEC"

Add DS Data page:

This is a form that lets you edit, add or delete DS Data. The call to the registry happens when you click save. Clicking cancel will undo all local changes that have not been pushed to the registry. If you have Key data in the registry, this page with say" Warning, you cannot add DS Data"

Add Key Data page:

Similar to Add DS Data page, does not currently work, assume it's an issue with CF. We are investigation on the CF slack channel.

Validation and error handling:

There is basic validation on form fields locally and errors returned from the registrar when there's a failure. All errors are displayed at a page-level error alert. Success is displayed as a succ

## Setup

When testing DS Data forms, here are a few valid Digests that you can use for testing:
ec0bdd990b39feead889f0ba613db4adecb4adec
ec0bdd990b39feead889f0ba613db4adec0bdd99

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [x] Checked keyboard navigability
- [x] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [x] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [x] Pulled this branch locally and tested it
- [x] Reviewed this code and left comments
- [x] Checked that all code is adequately covered by tests
- [x] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [x] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [x] Chrome
  - [ ] Microsoft Edge
  - [x] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/templates/domain_dnssec.html
Comment (h-m-f-t) [inline src/registrar/templates/domain_dnssec.html:10]: Replace "website" with "domain".

File: src/registrar/templates/domain_dnssec.html
Comment (h-m-f-t) [inline src/registrar/templates/domain_dnssec.html:46]: Change to:

> It is strongly recommended that you only enable DNSSEC if you know how to set it up properly at your hosting service. If you make a mistake, it could cause your domain name to stop working.

**Review by h-m-f-t — DISMISSED** [review]
A few small copy changes.

**Comment by h-m-f-t** [comment]
My review counts for very little, especially on this one, so I'll just comment and say that the copy update LGTM. Thanks @dave-kennedy-ecs!

**Comment by gabydisarli** [comment]
@dave-kennedy-ecs In the sidebar nav, can we rename `DNS name servers page` to just `Name servers` Now that the name servers page lives under a broader DNS section we don't need to use DNS again.

**Comment by gabydisarli** [comment]
Can we make the alert here span the width of the paragraph above?
Screenshot 2023-10-11 at 3 38 45 PM

File: src/registrar/templates/domain_dnssec.html
Comment (gabydisarli) [inline src/registrar/templates/domain_dnssec.html:44]: Not sure if this is the right place to add this but noticing that the alert has the wrong padding, should be 46 and I'm seeing 56 on inspection. Ignore if this is just a USWDS issue.

File: src/registrar/templates/domain_dnssec.html
Comment (gabydisarli) [inline src/registrar/templates/domain_dnssec.html:54]: space between enable DNSSEC button and preceding paragraph should be 24

Comment (gabydisarli) [reply src/registrar/templates/domain_dnssec.html:54]: Screenshot 2023-10-11 at 3 52 17 PM

**Comment by gabydisarli** [comment]
Please change padding between outline buttons and link button to 24 (leave the padding between outline buttons as is)
Screenshot 2023-10-11 at 4 00 22 PM

**Comment by gabydisarli** [comment]
For consistency, let's change this title back to just DNSSEC
Screenshot 2023-10-11 at 4 06 53 PM

**Comment by gabydisarli** [comment]
Screenshot 2023-10-11 at 4 11 23 PM
There's been some changes to this page. [See Figma link.](https://www.figma.com/file/aJbuDLJNNikqQObTuCNKQa/GetGov_Designs_ECS-Truss?type=design&node-id=1250%3A66410&mode=design&t=rNFinXKa1RAHD1jk-1)
- Copy updates
- Change default button to the add link style button (same used in name servers page)
- Add info alert (@rachidatecs please review this to confirm it would work for that edge case)

**Comment by gabydisarli** [comment]
Screenshot 2023-10-11 at 4 25 25 PM
Can we change this margin to 40?

File: src/registrar/assets/js/get-gov.js
Comment (zandercymatics) [inline src/registrar/assets/js/get-gov.js:279]: Why Num2?

**Comment by abroddrick** [comment]
More for @gabydisarli 
image
Key tag is sentence case but Digest Type is cased like a proper noun/ Title. Can we chose to use one or the other?

**Comment by abroddrick** [comment]
Also I noticed if you submit duplicates it rightly doesn't throw a registry error. But in version 2, should it mention "you submitted duplicate entries, only one was kept". More for @gabydisarli  but also the devs (and not something that should be included in this PR anyways as it's not a must)

**Comment by abroddrick** [comment]
Last thing, I feel iffy about the drop down for algorithm. @gabydisarli is this something we were planning to go over with cloudlfare? There are more algorithm types so I wonder if it is bad to not include all options

File: src/registrar/templates/domain_dnssec.html
Comment (rachidatecs) [reply src/registrar/templates/domain_dnssec.html:44]: The USWDS slim alert with icon has a padding left of 3.5em

Screen Shot 2023-10-13 at 11 37 14 AM

Overriding it will cause issues and require more refactoring. Ignoring.

Screen Shot 2023-10-13 at 11 35 53 AM

**Comment by rachidatecs** [comment]
> Please change padding between outline buttons and link button to 24 (leave the padding between outline buttons as is) Screenshot 2023-10-11 at 4 00 22 PM

We'll have to override the margins on these cancel buttons (or unstyled buttons) wherever we might use them in the future.

This extra margin also affects the stacked mobile view, is this desirable or should we apply the margin on tablet and up only:

Screen Shot 2023-10-13 at 11 45 40 AM
Screen Shot 2023-10-13 at 11 45 54 AM

**Comment by rachidatecs** [comment]
> Last thing, I feel iffy about the drop down for algorithm. @gabydisarli is this something we were planning to go over with cloudlfare? There are more algorithm types so I wonder if it is bad to not include all options
> 
> (devs, I am still reviewing so sorry this one takes a bit)

For context, we defined the literal for algorithm choices based on https://www.iana.org/assignments/dns-sec-alg-numbers/dns-sec-alg-numbers.xhtml

**Comment by dave-kennedy-ecs** [comment]
> Also I noticed if you submit duplicates it rightly doesn't throw a registry error. But in version 2, should it mention "you submitted duplicate entries, only one was kept". More for @gabydisarli but also the devs (and not something that should be included in this PR anyways as it's not a must)

It doesn't throw a RegistryError.  But Cloudflare only saves one of the duplicates.  And when we populate the cache and then reload the form after submission, the duplicate has been removed.  Based on your comments, I think we are fine for this PR.  But also this is kind of tricky to handle UI-wise, unless we error duplicates and not allow submission until resolved by the user.

**Comment by abroddrick** [comment]
> It doesn't throw a RegistryError. But Cloudflare only saves one of the duplicates. And when we populate the cache and then reload the form after submission, the duplicate has been removed. Based on your comments, I think we are fine for this PR. But also this is kind of tricky to handle UI-wise, unless we error duplicates and not allow submission until resolved by the user.

Yeah you're right the easiest way would be to have an error that blocks the user from moving forward if there is a duplicate, which wouldn't be clean and is unnecessary if epp removes it anyways. Would really just be something for @gabydisarli to be aware of as we get into user testing more and more, to see if it would even be worth enhancing this in the future

File: src/registrar/tests/common.py
Comment (abroddrick) [inline src/registrar/tests/common.py:861]: **comment** I do feel this makes it harder to set conditionals for a given domain to return a mock value based on call count or other conditionals.

File: src/registrar/views/domain.py
Comment (abroddrick) [inline src/registrar/views/domain.py:159]: **nitpick** eye twitch here too at the added line. feel free to ignore this comment

File: src/registrar/views/domain.py
Comment (abroddrick) [inline src/registrar/views/domain.py:245]: same here

File: src/registrar/views/domain.py
Comment (abroddrick) [inline src/registrar/views/domain.py:253]: i think this is on the wrong line and just needs to be moved up

File: src/registrar/views/domain.py
Comment (abroddrick) [inline src/registrar/views/domain.py:284]: **non-blocking:** I think we want to confirm this error message in the next ticket with @gabydisarli because I think we would want language to "try again later" or something to indicate what the user should do given an error occurred.

File: src/registrar/views/domain.py
Comment (abroddrick) [inline src/registrar/views/domain.py:298]: **non-blocking**: 👀+twitch at this empty line, just complaining feel free to ignore

File: src/registrar/views/domain.py
Comment (abroddrick) [inline src/registrar/views/domain.py:314]: **comment** approve of the use of "moot" in a PR

File: src/registrar/views/domain.py
Comment (abroddrick) [inline src/registrar/views/domain.py:297]:
```suggestion
class DomainDsDataView(DomainPermissionView, FormMixin):
```
Total suggestion, just looks like the D should be capitalized here

File: src/registrar/views/domain.py
Comment (abroddrick) [inline src/registrar/views/domain.py:379]:
```suggestion
        """The formset is valid, save DsData""
```

**suggestion**

File: src/registrar/views/domain.py
Comment (abroddrick) [inline src/registrar/views/domain.py:400]: fyi given convo with @dave-kennedy-ecs I think this also triggers a reinit of the domain object, correct Dave? may be worth editing here or in the follow on ticket for removing key data. 

**non-blocking callout**

File: src/registrar/views/domain.py
Comment (abroddrick) [inline src/registrar/views/domain.py:403]: We do need error handling to follow what we are going to do in Nameservers and security email (currently being worked on by Rachid). In those cases we have 1.) a case where registry just can't connect and we display a can't connect message 2.) a case where we can't update for other reasons and display a generic error. 

@rachidatecs perhaps we add this to the keydata ticket? Or would you prefer another ticket be made to address it. 

Don't block this ticket for that though as that seems unnecessary given the follow on ticket, but felt I should call it out while it's obviously

File: src/registrar/views/domain.py
Comment (abroddrick) [inline src/registrar/views/domain.py:406]: **comment** actualy like that you log both here, nice idea!

**Review by abroddrick — APPROVED** [review]
No blocking comments, tested and reviewed code and happy with it as is. Please see minor comments/ suggestions before merging

File: src/registrar/tests/common.py
Comment (dave-kennedy-ecs) [reply src/registrar/tests/common.py:861]: linter made us do it ...

File: src/registrar/views/domain.py
Comment (dave-kennedy-ecs) [reply src/registrar/views/domain.py:298]: I removed a bunch of empty lines, to hopefully relieve some of the eye twitches

File: src/registrar/views/domain.py
Comment (dave-kennedy-ecs) [reply src/registrar/views/domain.py:398]: done

File: src/registrar/forms/common.py
Comment (dave-kennedy-ecs) [reply src/registrar/forms/common.py:25]: was not able to successfully write a digest to cloudflare for a digest type other than 1.  other documents suggested that 1 is the only valid option.  will need cloudflare to tell us what are valid values to use here.  this is our best guess based on current documentation and testing.

File: src/registrar/views/domain.py
Comment (dave-kennedy-ecs) [reply src/registrar/views/domain.py:403]: thanks

File: src/registrar/views/domain.py
Comment (dave-kennedy-ecs) [reply src/registrar/views/domain.py:397]: thank you.  merging this before the caching ticket, so will fix this as well in the caching ticket, where there is a means to handle it properly

File: src/registrar/assets/js/get-gov.js
Comment (dave-kennedy-ecs) [reply src/registrar/assets/js/get-gov.js:342]: added a comment block on this one

**Comment by rachidatecs** [comment]
UPDATE:

This PR is no linger canonical after refactoring the JS in the Namerservers work. To implement Key data again, follow this PR in reverse then follow the commit: https://github.com/cisagov/getgov/commit/6d3f5bf5ef9d4c529feba90313608020e5df834b

**Pull request merged**
